[scripts/get_cmake_vars][scripts/ports.cmake] fix CMake4 on OSX

strega-nil-re opened the pull request:
It looks like CMake 4 may not always initialise CMAKE_OSX_SYSROOT, and `get_cmake_vars`/`z_vcpkg_get_cmake_vars` would fail if that happened. I simply fixed it the same way that the port `vcpkg_cmake_get_vars` does.

This additionally sets the environment variable `CMAKE_POLICY_VERSION_MINIMUM` to `4.0` when using CMake 4.0, so that any ports that use an ancient `cmake_minimum_required` are taken care of.

Fixes microsoft/vcpkg#44697

strega-nil committed:
[scripts/get_cmake_vars] fix CMake4 on OSX

It looks like CMake 4 may not always initialise CMAKE_OSX_SYSROOT, and
`get_cmake_vars`/`z_vcpkg_get_cmake_vars` would fail if that happened.
I simply fixed it the same way that the port `vcpkg_cmake_get_vars`
does.

Fixes microsoft/vcpkg#44697

strega-nil-re commented:
@microsoft-github-policy-service agree company="reMarkable AS"

strega-nil-re commented:
@microsoft-github-policy-service terminate

strega-nil-re commented:
(I think I accidentally agreed on this account for myself, rather than as a paid contributor from reMarkable, so redoing the agreement)

Cheney-W commented:
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#update-the-port-version-field-in-the-manifest-file-of-any-modified-ports
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#update-the-version-files-in-versions-of-any-modified-ports

Neumann-A commented on ports/crc32c/fix_cmake4.patch:
```diff
@@ -0,0 +1,19 @@
+ # found in the LICENSE file. See the AUTHORS file for names of contributors.
+ 
+-cmake_minimum_required(VERSION 3.1)
++cmake_minimum_required(VERSION 3.1...4.0)
```
I mean you could add all those patches but it is probably better to use:
`set(ENV{CMAKE_POLICY_VERSION_MINIMUM} 3.5)`
in `ports.cmake` if the `CMAKE_VERSION` is bigger/equal to 4

strega-nil-re replied:
I really dislike this choice, because it means that this will break again when CMake drops support for 3.5, and also generally means we're not making sure that specific ports work. However, if the vcpkg team thinks this is a better choice, I'm also happy to do that.

Neumann-A replied:
> I really dislike this choice, because it means that this will break again when CMake drops support for 3.5

How is: `cmake_minimum_required(VERSION 3.1...4.0)` better in this regard? 
It will break when CMake will drop support for 4.0, the time span is probably just longer.

strega-nil-re replied:
It says "I have actually tested this code with 4.0, and it works".

strega-nil committed:
[scripts/ports.cmake] set CMAKE_POLICY_VERSION_MINIMUM

CMake 4 broke compatiblity with CMake pre-3.5, so we need to set this variable
to avoid ports which have an earlier cmake_minimum_required(VERSION) breaking.

Cheney-W commented:
The failure of `proxygen:x64-linux` is unrelated to this PR.

ADKaster commented:
It looks like this PR would also partially address https://github.com/microsoft/vcpkg/issues/44726 ? At least, the second commit will hide the error from vcpkg users.

ADKaster commented on scripts/ports.cmake:
```diff
@@ -3,6 +3,9 @@ if(POLICY CMP0174)
     # Use CMake 3.31 behavior for cmake_parse_arguments(PARSE_ARGV)
     cmake_policy(SET CMP0174 NEW)
 endif()
+if(CMAKE_VERSION GREATER_EQUAL "4.0")
```
shouldn't this be `VERSION_GREATER_EQUAL` ?

https://cmake.org/cmake/help/latest/variable/CMAKE_VERSION.html

> Use the [if()](https://cmake.org/cmake/help/latest/command/if.html#command:if) command VERSION_LESS, VERSION_GREATER, VERSION_EQUAL, VERSION_LESS_EQUAL, or VERSION_GREATER_EQUAL operators to compare version string values against CMAKE_VERSION using a component-wise test. Version component values may be 10 or larger so do not attempt to compare version strings as floating-point numbers.

BillyONeal commented:
proxygen:x64-linux is a baseline issue: https://dev.azure.com/vcpkg/public/_build/results?buildId=114048&view=logs&j=f79cfdd7-47a8-597f-8f57-dc3e21a8f2ad&t=73c5513e-c47c-57a4-379a-581f7e429f13

BillyONeal commented on scripts/ports.cmake:
```diff
@@ -3,6 +3,9 @@ if(POLICY CMP0174)
     # Use CMake 3.31 behavior for cmake_parse_arguments(PARSE_ARGV)
     cmake_policy(SET CMP0174 NEW)
 endif()
+if(CMAKE_VERSION GREATER_EQUAL "4.0")
+    set(ENV{CMAKE_POLICY_VERSION_MINIMUM} 3.5)
+endif()
```
```suggestion
if(CMAKE_VERSION VERSION_GREATER_EQUAL "4.0")
    set(ENV{CMAKE_POLICY_VERSION_MINIMUM} 3.5)
endif()
```

BillyONeal committed:
Use VERSION_GREATER_EQUAL

BillyONeal commented:
Thanks for the fix! Related: #44273 has other fixes for 4.x but I think we should take this more targeted change to put out the current fire drill.
